nchelpers/__init__.py

```diff
@@ -1769,6 +1769,21 @@ def climo_periods(self):
         """List of the standard climatological periods (see function
         ``standard_climo_periods``) that are a subset of the date range in
         the file."""
+
+        def final_climatology_interval(e):
+            """Returns the first day of the final interval (individual month,
+            season, etc) needed for this climatology.
+            To verify that data is present to calculate the climatology,
+            a timestamp during or after this interval is needed (>= the date
+            returned by this function).
+            Seasonal datasets require special handling, as their climatologies end
+            in autumn, not winter, so they don't go through to the end of the year."""
+            if self.time_resolution == 'seasonal' and e.month == 12:
```

Comment:
This is probably a failure of understanding on my part, but you have a special case for `seasonal` both here and in `truncate_to_resolution`. Is that (a) correct, and (b) necessary? If you could handle the special case in just one place it would be easier to understand, I think.

Reply:
OTOH, it clearly works.

Reply:
I view these as being two separate special cases, even though they both handle weirdness relating to "winter" spanning the year boundary. But perhaps there's a unifying framework I'm overlooking; I'd love to hear about some elegant approach I missed!

1. The special case handling in `date_utils.py` handles a special case relating to _how the interval comprising "winter" is defined_, specifically what its starting timestamp is. 
If we didn't have the special case and winter dates followed the default behaviour, dates in January or February would be (incorrectly) assigned to an interval whose first timestamp was the invalid month 0; dates in December would (correctly) be assigned to an interval whose first timestamp was December 1.
Basically the special case says "if the month is January or February, its season began the previous year"

2. The special case handling in `nchelpers/__init.py__` handles a special case relating to _how climatologies of seasonal data are defined_ and what values should be included in the calculation. The way we have been calculating climatologies, a climatology that runs from year X to year Y does include the winter value for the interval running from December of X-1 to February of X, but it does not include the winter value for the interval running from December of Y to February of Y+1.
If we didn't have the special case and followed the default, a climatology from year X to year Y would include both the winters from December of X-1 to February of X and the winter from December of Y to February of Y+1. There's definitely an argument to be made that this would be a reasonable and valid way to calculate the seasonal climatology, but 1) it's not the way we've been doing it (... I think), and 2) the HadGEM2 datasets that have caused all this trouble don't extend far enough to allow defining a seasonal climatology this way.
Basically the special case says "even though the bounds of this climatology include December of the final year Y, the value representing December of Y to February of Y+1 is not needed to construct a climatology"

Reply:
I see; it was my failure of understanding here. Given that, maybe add these discussions (or some abbreviated version of them) as comments in each place?